TaskTree Implementation

MMachadoUofT opened the pull request:
I've created the TaskTree methods, which were probably the most "computationally difficult" methods to write, meaning that I should be able to finish all of the backed now by the end of the week and divert focus onto my create curriculum use case. Time permitting, I may also assign myself the work of creating the program Driver if I'm able to complete that in a timely manner. 

I also want to mention a few things; from the perspective of the use case interactors, I don't know how helpful it would be to actually have methods that find TaskTree objects based on Task, as it is the Task itself we are trying to find when accessing a tree (hence the "ByID" methods, which I think are far more relevant). By their very nature, methods that require an entity as a parameter that aren't inherently about setting that entity to something seem a bit useless, considering the information the interactors get from the controllers is never an entity. Furthermore, I think that an exception along the lines of "NullTaskException," or "NoTaskAssigned" would be useful, just in case. In practice, we'd never want to have a TaskTree with no Task, as the point of the TaskTree structure is simply organizational, but if anyone would like me to do this, I would be glad to. For the sake of speed and progress, however, I may make this a separate Feature/Mini-feature/Subissue and label it low-priority, just to make sure the functionality of our entire program is set before we begin trying to manage all of our edge cases.

As a last note, I definitely think we need a little bit of refactoring with regards to the packaging we have. Currently, we have a use_cases package with only one use case, and Baylee's is simply on the top layer. This might've just been an issue with IntelliJ not liking empty packages, but if someone wants to make a refactoring PR that would awesome.

MMachado05 committed:
Created attributes and constructors for the CommonTaskTree class

I've also made TaskTree extend the Serializable interface for data persistence purposes.

MMachado05 committed:
Created constructors and attributes for the CommonTaskTree class

I've also added a testing class, but I'm still setting up JUnit so the tests there don't work. I'll make another commit once I've gotten it working

MMachado05 committed:
Merge branch 'main' into 11-subissue-13-tasktree-classes

MMachado05 committed:
Created subTaskTree addition/removal methods

- addSubTaskTree
- removeSubTaskTree
- removeChildTaskTree
The last of which contains a recursive call to removeSubTaskTree

MMachado05 committed:
Refactored entityLayer packages to entity_layer

This change was pushed to main, but just to be safe I did it on this local branch, as well.

MMachado05 committed:
Created TaskTreeFactory classes

I've also implemented them, as they're relatively easy to do. Most of the actual "creation" of a TaskTree is done by its methods, anyway

MMachado05 committed:
Implemented getChildTaskTree methods

They're rather similar, and frankly I don't see the "ByTask" being used as it requires an actual Task instance. I suppose the implementation I had in my head had one depending on the other, but that seems silly now. I may refactor this later and remove the "ByTask" method entirely. I'd like a second opinion when I make my PR.

MMachado05 committed:
Implemented toList methods

One thing that might be useful, especially for the purposes of presenting the Curriculum's TaskTree to the UI, would be maybe doing a hashmap where each key is a supertask ID, and each value is a Task object. I'll make a note of this and include it in my PR to bring attention to it.

MMachado05 committed:
Implemented completeTask and updateTask

These are coupled, as completing a Task necessarily involves making sure that all of its superTask's completions represent the partial completion of its subtasks. The actual algorithm is rather inefficient I'm now realizing, but I'll worry about optimization later.

MMachado05 committed:
Implemented CommonTaskTree getters and setters

MMachado05 committed:
Created and implemented assorted TaskTree tests

I'm making this commit primarily to change something else in the CommonTaskTree code, which, being unrelated to this testing code as of now, will be in a separate commit. So far, all tests have passed.

MMachado05 committed:
Included a call to updateTask in addSubTaskTree

This also made me realize that an exception for a Task-less TaskTree may be useful, so I'll make a note of that in my PR.

MMachado05 committed:
Included a call to updateTask in addSubTaskTree (2)

I committed that too early. I like to write my commit messages in IntelliJ before I'm done just so that I know what I thought of. Oops.

MMachado05 committed:
Merge branch 'main' into 11-subissue-13-tasktree-classes

# Conflicts:
#	src/main/java/entity_layer/CommonTaskTree.java
#	src/main/java/entity_layer/User.java

MMachado05 committed:
Moved UserFactory classes into entity_factories

Hopefully this doesn't cause any unwanted code conflicts

MMachado05 committed:
Created more TaskTree tests; fixed a bug in updateTask

More importantly, there is now a small if statement that simply skips the updateTask() method if there is no assigned Task. I would like for this instead to return an exception, maybe TasklessTreeError or something. I also fixed a small bug in removeChildTaskTree() that the warnings made me learn about.

MMachado05 committed:
Added and passed final tests for CommonTaskTree

I've made a couple of new tests, namely some regarding testing the state of a taskTree (has subtasks, supertask, etc), and with it my code coverage is 100% for the methods in the class. I'll be making my PR with this.

MMachado05 committed:
Removed unnecessary comments and refactored addSubTaskTree

Bhan-Aayush commented:
looks good to me

MMachado05 committed:
Merge branch 'main' into 11-subissue-13-tasktree-classes

# Conflicts:
#	src/main/java/UI/LoginUI.java
#	src/test/java/UserRegistrationTests.java

dadler83 reviewed (dismissed):
Looks good to me! As long as it passes the tests and behaves as expected

MMachado05 committed:
Merge branch 'main' into 11-subissue-13-tasktree-classes

# Conflicts:
#	src/main/java/entity_factories/CommonUserFactory.java

bmguiler reviewed (dismissed):
looks good, and thanks for adding tests. now my use case will work properly - yay everything is coming together

MMachado05 committed:
Light refactoring for UserFactory

After updating my branch, there were some errors with imports as a result of some classes being in the wrong place. I moved them into the proper packages, then made sure the tests still worked after their import statements were appropriately adjusted.

Bhan-Aayush commented:
Looks good to me this will be helpful for my tasktree UI.

dadler83 commented on src/main/java/entity_factories/CommonTaskTreeFactory.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public TaskTree create() {
+        return new CommonTaskTree();
+    }
```
there might be utility in creating a tasktree with a parent, like having the create method take an id, and make a subtask tree of that parent

dadler83 commented on src/main/java/entity_layer/CommonTaskTree.java:
```diff
@@ -1,24 +1,62 @@
+
+        if (oldParent != null) {
+            oldParent.removeSubTaskTree(taskTree);
+            oldParent.updateTask();
```
if you remove the last TaskTree belonging to a parent, you will get a divide by 0 error from the updateTask method because you don't account for the case where the size of the ArrayList is 0

MMachadoUofT replied:
Didn't catch that! Thank you

dadler83 commented on src/main/java/entity_layer/CommonTaskTree.java:
```diff
@@ -30,7 +68,12 @@ public void addSubTaskTree(TaskTree taskTree) {
     public boolean removeSubTaskTree(TaskTree taskTree) {
-        return false;
+        if (this.subTaskTrees.contains(taskTree)) {
+            this.subTaskTrees.remove(taskTree);
```
I think this could cause issues for completion percentages because after a subtask tree is removed, there is no update on the completion percentages; your test cases should show you this

MMachadoUofT replied:
I forgot to do implement this! Good catch, I'll write a case for it now, too.

dadler83 commented on src/main/java/entity_layer/CommonTaskTree.java:
```diff
@@ -42,7 +85,17 @@ public boolean removeSubTaskTree(TaskTree taskTree) {
     public boolean removeChildTaskTree(TaskTree taskTree) {
-        return false;
+        // Check if it's a direct subtree
+        if (this.removeSubTaskTree(taskTree))
```
From this implementation, it's clear to me that the removeSubTaskTree method should be private because if you try this method, it will always call it; this functionally eliminates the need for someone outside of this class to ever use the removeSubTaskTree method because it is a less powerful helper method

dadler83 replied:
Also, there should probably be a way to search out and remove a TaskTree by id

MMachadoUofT replied:
Very true, I hadn't caught that. I'll change it now.

MMachadoUofT replied:
As for the ID bit, My thought is that the existing methods would allow an interactor to do that already. However, I may add this as an actual method later down the line and expand the entities capabilities.

dadler83 commented on src/main/java/entity_layer/CommonTaskTree.java:
```diff
@@ -54,7 +107,19 @@ public boolean removeChildTaskTree(TaskTree taskTree) {
     public TaskTree getChildTaskTreeByTask(Task task) {
-        return null;
+        // Check if this is the matching tree
+        if (this.task == task)
```
This implementation might be problematic because '==' works differently on objects in Java. This will only return true if the given task and the compared have the same address in memory. You might need to use .equals() and overwrite it for tasks.
https://www.baeldung.com/java-comparing-objects

MMachadoUofT replied:
That's more or less how I wanted it to be implemented, in fact. To compare the actual Task objects in memory. However, looking at it now, I may forego these methods entirely, as they really aren't going to be useful for the interactors since they are only given the taskID's.

dadler83 reviewed (commented):
Generally, very good work; I think there are some edge cases that have not been tested, and I also think that we need to think about how to get this information to the UI -- because the format of the tree structure is super important in terms of display, but I don't see the toList methods being useful because they completely eliminate the structural component of the data they return

MMachado05 committed:
Make removeSubTaskTree() private

As it stands, the removeChildTaskTree() method is far more powerful and can accomplish the same thing if desired. It works better to have removeSubTaskTree as a private helper method just for removeChildTaskTree.

MMachado05 committed:
Fixed divide-by-zero bug in updateTask

I was notified that my implementation, as it was, would divide by zero if a TaskTree had its last task removed. I've included both a call to updateTask in the removeSubTaskTree method to assure TaskTrees are always updated, and a setting of an ignoring of the taskCount in the case that the last TaskTree was just removed. If the last TaskTree is removed, then at that point, the parent Tree would already have completion set only either to 0 or 100, so no adjustment of completion need be made.

MMachado05 committed:
Created and implemented resetTask() method

MMachadoUofT commented:
One thing I might consider is entirely removing updateTask() from internal calls. That is, the tree will only worry about going _down_ the line, updating its children, and then the use case interactor can call updateTask(). There's a lot of unnecessary calls to updateTask() going on in the current implementation, and, while it works perfectly fine, is definitely a bit underperforming. Plus, this might help stick to clean architecture a little bit, as the interactors will be given more appropriate responsibility in modifying existing entities

bmguiler reviewed (dismissed):
Looks good, I am just wondering if all of these methods will be necessary? I would hate for you to write all of these and some of them never get used.

dadler83 reviewed (dismissed):
It looks like the issues I spoke of are fixed; however, you should remove some of these comments, like the one about optimization under updateTask() -- if you are going to optimize it do so, but a recursive method by nature requires many repeat calls. As for empty TaskTrees with no tasks, you should probably raise an exception, maybe a custom one, since there should never be a TaskTree with no associated Task. It's probably best if you raise another PR and ticket for cleaning this stuff up.

MMachado05 committed:
Merge branch 'main' into 11-subissue-13-tasktree-classes

MMachado05 committed:
Changed completeTaskUseCaseTest imports to work with refactored packaging

Bhan-Aayush reviewed (approved):
Great work! However, how do we want the format of the tree structure related to this as we need to figure it out on the display? You have used clean architecture I see, however for next time I think adding more comments would be helpful for other people to understand the code. (Not necessary but might be helpful for next time, as I see a lot of changes in the comments)
